client/service/src/metrics.rs

```diff
@@ -0,0 +1,379 @@
+
+use crate::NetworkStatus;
+use netstat2::{TcpState, ProtocolSocketInfo, iterate_sockets_info, AddressFamilyFlags, ProtocolFlags};
+use prometheus_endpoint::{register, Gauge, U64, F64, Registry, PrometheusError, Opts, GaugeVec};
+use sc_client::ClientInfo;
+use sc_telemetry::{telemetry, SUBSTRATE_INFO};
+use std::convert::TryFrom;
+use sp_runtime::traits::{NumberFor, Block, SaturatedConversion, UniqueSaturatedInto};
+use sp_transaction_pool::PoolStatus;
+use sp_utils::metrics::register_globals;
+use sysinfo::{ProcessExt, System, SystemExt};
+
+#[cfg(not(unix))]
+use sysinfo::get_current_pid;
+
+#[cfg(unix)]
+use procfs;
+
+struct PrometheusMetrics {
+	// system
+	load_avg: GaugeVec<F64>,
+
+	// process
+	cpu_usage_percentage: Gauge<F64>,
+	memory_usage_bytes: Gauge<U64>,
+	netstat: GaugeVec<U64>,
+	threads: Gauge<U64>,
+	open_files: GaugeVec<U64>,
+
+	// -- inner counters
+	// generic info
+	chain_info: GaugeVec<U64>,
+	ready_transactions_number: Gauge<U64>,
+
+	// I/O
+	network_per_sec_bytes: GaugeVec<U64>,
+	database_cache: Gauge<U64>,
+	state_cache: Gauge<U64>,
+	state_db: GaugeVec<U64>,
+}
+
+impl PrometheusMetrics {
+	fn setup(registry: &Registry, name: &str, version: &str, roles: u64)
+		-> Result<Self, PrometheusError>
+	{
+        register(Gauge::<U64>::with_opts(
+            Opts::new(
+                "build_info",
+                "A metric with a constant '1' value labeled by name, version"
+            )
+                .const_label("name", name)
+                .const_label("version", version)
+		)?, &registry)?.set(1);
+		
+        register(Gauge::<U64>::new(
+            "node_roles", "The roles the node is running as",
+		)?, &registry)?.set(roles);
+
+		register_globals(registry)?;
+		
+		Ok(Self {
+
+			// system
+			load_avg: register(GaugeVec::new(
+				Opts::new("load_avg", "System load average"),
+				&["over"]
+			)?, registry)?,
+
+			// processhttps://github.com/paritytech/substrate/issues/5485
+			memory_usage_bytes: register(Gauge::new(
+				"memory_usage_bytes", "Node memory usage",
+			)?, registry)?,
+
+			cpu_usage_percentage: register(Gauge::new(
+				"cpu_usage_percentage", "Node CPU usage",
+			)?, registry)?,
+
+			netstat: register(GaugeVec::new(
+				Opts::new("netstat_tcp", "Current TCP connections "),
+				&["status"]
+			)?, registry)?,
+
+			threads: register(Gauge::new(
+				"threads", "Number of threads used by the process",
+			)?, registry)?,
+
+			open_files: register(GaugeVec::new(
+				Opts::new("open_file_handles", "hold by the process"),
+				&["fd_type"]
+			)?, registry)?,
+
+			// --- internal
+
+			// generic internals
+
+			chain_info: register(GaugeVec::new(
+				Opts::new("chain_info", "Height and Fork Info of the chain"),
```

Comment:
The `_info` metric implies that a metric is static in the Prometheus ecosystem.

Do I understand correctly that you renamed `block_height` to `chain_info` to incorporate the number of leaves? If so I would suggest introducing a new metric for the number of leaves as block height should be ever increasing and number of leaves floating within some range. What do you think?

Reply:
Is there any documentation on what kind of metrics makes sense to have bundled together?

I really don't care, but I considered all that "informational" but related. you will probably not see any forks until the best-block-height is your block-height (because we immediately discard forks that didn't turn out to exist). So yeah, maybe these are unrelated.

Reply:
> Is there any documentation on what kind of metrics makes sense to have bundled together?

Yes. I think a very good resource and guideline is https://prometheus.io/docs/practices/naming/

Next to the resource above, I also like the thought to use labels to differentiate things that could possibly be used in arithmetic operations. E.g. it makes sense to subtract `finality` from `best` to see how far finality is behind, but as far as I can tell it does not make sense to substract the number of leaves from any of the former. Please correct me if I am missing something.